File: sdk/python/src/aether/evaluate.py
from __future__ import annotations

import time
from dataclasses import dataclass

from .corpus import load_golden
from .engine import Engine
from .plan import classify_query


@dataclass
class EvalExampleResult:
    id: str
    question: str
    hit: bool
    reciprocal_rank: float
    recall_at_5: float
    predicted_sources: list[str]
    path: str
    latency_ms: int


@dataclass
class EvalReport:
    results: list[EvalExampleResult]
    recall_at_5: float
    mrr: float
    mean_latency_ms: float

    def __str__(self) -> str:
        hits = sum(1 for r in self.results if r.hit)
        return f"Recall@k {self.recall_at_5:.0%} · MRR {self.mrr:.2f} · {hits}/{len(self.results)} hits · {self.mean_latency_ms:.0f} ms"


def run_golden_eval(engine: Engine | None = None) -> EvalReport:
    eng = engine or Engine.northstar()
    results: list[EvalExampleResult] = []
    for ex in load_golden():
        t0 = time.time()
        hits = eng.search(ex["question"], 8)
        predicted: list[str] = []
        for h in hits:
            if h.document_id not in predicted:
                predicted.append(h.document_id)
        ranks = [predicted.index(s) for s in ex["expectedSources"] if s in predicted]
        best = min(ranks) if ranks else -1
        hit = best >= 0
        rr = (1 / (best + 1)) if hit else 0.0
        hits_at = sum(1 for s in ex["expectedSources"] if s in predicted[:8])
        recall = hits_at / len(ex["expectedSources"]) if ex["expectedSources"] else 0.0
        results.append(
            EvalExampleResult(
                id=ex["id"],
                question=ex["question"],
                hit=hit,
                reciprocal_rank=rr,
                recall_at_5=recall,
                predicted_sources=predicted,
                path=classify_query(ex["question"]).path,
                latency_ms=int((time.time() - t0) * 1000),
            )
        )
    n = len(results) or 1
    return EvalReport(
        results=results,
        recall_at_5=sum(r.recall_at_5 for r in results) / n,
        mrr=sum(r.reciprocal_rank for r in results) / n,
        mean_latency_ms=sum(r.latency_ms for r in results) / n,
    )

File: src/lib/rag/evaluate.ts
import { GOLDEN_EVAL } from "./corpus";
import { searchEngine } from "./engine";
import { classifyQuery } from "./plan";
import type { EvalRunResult } from "./types";

export async function runGoldenEval(): Promise<{
  results: EvalRunResult[];
  recallAt5: number;
  mrr: number;
  meanLatency: number;
}> {
  const results: EvalRunResult[] = [];
  for (const ex of GOLDEN_EVAL) {
    const t0 = Date.now();
    const hits = searchEngine(ex.question, 8);
    const predicted = unique(hits.map((h) => h.documentId));
    const ranks = ex.expectedSources.map((src) => predicted.indexOf(src)).filter((i) => i >= 0);
    const best = ranks.length ? Math.min(...ranks) : -1;
    const hit = best >= 0;
    const reciprocalRank = hit ? 1 / (best + 1) : 0;
    const hitsAt5 = ex.expectedSources.filter((s) => predicted.slice(0, 8).includes(s)).length;
    const recallAt5 = ex.expectedSources.length ? hitsAt5 / ex.expectedSources.length : 0;
    results.push({
      id: ex.id,
      question: ex.question,
      kind: ex.kind,
      predictedSources: predicted,
      hit,
      reciprocalRank,
      recallAt5,
      path: classifyQuery(ex.question).path,
      latencyMs: Date.now() - t0,
      confidence: hits[0]?.score ?? 0,
    });
  }
  const recallAt5 = mean(results.map((r) => r.recallAt5));
  const mrr = mean(results.map((r) => r.reciprocalRank));
  const meanLatency = mean(results.map((r) => r.latencyMs));
  return { results, recallAt5, mrr, meanLatency };
}

function unique(xs: string[]): string[] {
  return [...new Set(xs)];
}

function mean(xs: number[]): number {
  if (!xs.length) return 0;
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

export { GOLDEN_EVAL };
